### ISSUE TITLE
Stuck on "Retrieving character list" after logout with timer

### ISSUE THREAD

**Issue by Lankytim** [cmangos/issues#945]
**Current behavior**: 
Character selection screen stuck on "Retrieving character list" after character logout.  Need to hit Cancel then re-select realm to get character list back.  Also sometimes server crashes on re-login.

**Expected behavior**: 
Character selection screen should appear with characters populated, ready to re-log in to game.

**Steps to reproduce**:
1. Log in with an account containing at least 2 characters
2. Move character A into a safe zone (Green map or an Inn) then choose Logout from game menu
3. Move character B to a Yellow map (or somewhere that causes the 20sec logout timer) then choose Logout from game menu and wait
4. Log in immediately with character B again - can cause server crash (1 time in 4 with my testing); or log in with character A again and everything seems fine

**Client version**: 
2.4.3 (8606)

**Commit hash**: 
https://github.com/cmangos/mangos-tbc/commit/69bf156f8956e4bf7c4421f1d5e81835944af161

**Database version**: 
TBC-DB 1.5.0
ACID 2.4.3

**Operating system**: 
Linux Debian 8 Jessie 64bit

[//]: # (If this is a crash report, include the crashlog from a debug build with https://gist.github.com/)

**Comment by katz1** [comment]
I can confirm the part about not being able to return to the character selection screen after the 20 second logout. Tested with classic core, both Windows and Debian.

After you press cancel on the 'Retrieving character list', you will get a 'Session Expired'.

**Comment by Lankytim** [comment]
I found the "InstantLogout" setting in mangosd.conf.

The behavior I described above occurs with InstantLogout = 1 (default)

The issue is NOT present with InstantLogout = 0

**Comment by Lankytim** [comment]
I should have mentioned earlier - this is all with a GM level 0 account.

**Comment by Phatcat** [comment]
That could indicate that the issue lies with the InstantLogout functionality.

@cyberium it seems like the issues with the conf settings are piling up. I think we're gonna have to take a closer look at each and every one of them...

**Comment by katz1** [comment]
As I understand InstantLogout = 0 enables instant logout for all security levels. If this is the case the issue is not present with InstantLogout = 0, because there won't be a logout timer for anybody.

As both Lankytim and I stated, the issue happens when logging out with a timer. So the InstantLogout config value is unrelated to the issue itself in my opinion, or I may have misunderstood you guys horribly.

**Comment by Phatcat** [comment]
@katz1 Other way around.

**Comment by Lankytim** [comment]
I made the comment re: InstantLogout to at least show that a) the issue occurs with a default setting; b) setting it to 0 could perhaps be used as an interim workaround by affected server admins.

I gladly leave determining any relationship with the root cause up to the experts!  :)

**Comment by katz1** [comment]
> setting it to 0 could perhaps be used as an interim workaround by affected server admins

That is true, thanks for the tip. I changed the flag now as well.

@Phatcat 
> Other way around.

From the config:

>    InstantLogout
>        Enable or disable instant logout for security level (0..4) or high (NOT in combat/while dueling/while falling)
>        Default: 1 (Mods/GMs/Admins)

InstantLogout 0 enables instant logout for security level 0 and above, so there won't be a logout timer for anyone. I tested in on classic.

**Comment by Phatcat** [comment]
>Enable or disable instant logout for security level (0..4) ...

 Pretty sure it enables / disable instant logout for any security level or our commenting sure is a bit misleading.

**Comment by NocFenix** [comment]
I noticed this exact same issue as well. Digging into the code, I found a fix that seems to work. I haven't noticed any negative side effects, but, if you want to try it out:

In WorldSession.cpp around line 293:

    // check if we are safe to proceed with logout
    // logout procedure should happen only in World::UpdateSessions() method!!!
    if (updater.ProcessLogout())
    {
        ///- If necessary, log the player out
        const time_t currTime = time(nullptr);

        if (m_Socket->IsClosed() || (ShouldLogOut(currTime) && !m_playerLoading))
        {
            LogoutPlayer(true);
	        return true; // false;
        }
    }

Changed `return false;` to `return true;` From what I can tell, this change basically says to keep the session open. When it was closing the session it expired and logged the player out; thus preventing the character list from repopulating.

Again, I have not seen any adverse effects yet from this change, but I'll post back if I do find any.

**Comment by ulfgebhardt** [comment]
Can confirm the Problem on https://github.com/cmangos/mangos-tbc/tree/Last_mangos-tbc_PreBuild
InstantLogout = 1
Gm Level: 0/3

**Comment by Lankytim** [comment]
I tried for a while today but was unable to reproduce the crash that I originally mentioned.

When the crash I mentioned happened, I had very recently done a "make install" and run the DB installation script, both while the server was under a temporary VMware snapshot as part of my deployment method.  There was a chance that the disk I/O would be slightly slower at that time and maybe MySQL was busy doing something too.  I tried snapshotting as well as artificially starving the VM of CPU today in a hope to reproduce some race condition, but alas, she would not crash.

Have amended the issue title accordingly.
